File: Core/src/main/java/dev/lynith/core/ui/ComponentWithChildren.java
package dev.lynith.core.ui;

import dev.lynith.core.ui.callbacks.ComponentCallbacks;
import dev.lynith.core.ui.callbacks.impl.Destroyed;
import dev.lynith.core.ui.styles.AbstractComponentStyles;
import dev.lynith.core.versions.renderer.IRenderer;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public class ComponentWithChildren<C extends Component<C, S>, S extends AbstractComponentStyles<C, S>> extends Component<C, S> {

    @Getter @Accessors(fluent = false)
    private final CopyOnWriteArrayList<Component<?, ?>> children = new CopyOnWriteArrayList<>();

    @Override
    public void render(IRenderer ctx) {
        for (Component<?, ?> child : children) {
            child.render(ctx);
        }
        super.render(ctx);
    }

    @Override
    public void init() {
        listener(Destroyed.class, () -> getChildren().clear());

        for (Component<?, ?> child : children) {
            child.init();
        }

        super.init();
    }

    @Override
    public void update() {
        for (Component<?, ?> child : children) {
            child.update();
        }
        super.update();
    }

    @Override
    public void callCallbacks(Class<? extends ComponentCallbacks.CallbackInterface> callback, Object... args) {
        for (Component<?, ?> child : children) {
            child.callCallbacks(callback, args);
        }
        super.callCallbacks(callback, args);
    }

    public C add(Component<?, ?> child) {
        children.add(child);
        return (C) this;
    }

    public C remove(Component<?, ?> child) {
        children.remove(child);
        return (C) this;
    }

    public C remove(int index) {
        children.remove(index);
        return (C) this;
    }

    public C clear() {
        children.clear();
        return (C) this;
    }

    public C add(Component<?, ?>... children) {
        for (Component<?, ?> child : children) {
            add(child);
        }
        return (C) this;
    }

    public <E extends Component<?,?>> C add(List<E> children) {
        for (Component<?, ?> child : children) {
            add(child);
        }
        return (C) this;
    }

    public C remove(Component<?, ?>... children) {
        for (Component<?, ?> child : children) {
            remove(child);
        }
        return (C) this;
    }

    public <E extends Component<?,?>> C remove(List<E> children) {
        for (Component<?, ?> child : children) {
            remove(child);
        }
        return (C) this;
    }

    public C remove(int... indexes) {
        for (int index : indexes) {
            remove(index);
        }
        return (C) this;
    }

}

File: Core/src/main/kotlin/dev/lynith/core/ui/components/ComponentWithChildren.kt
package dev.lynith.core.ui.components

import dev.lynith.core.ui.callbacks.ComponentEvent
import dev.lynith.core.ui.callbacks.impl.Destroyed
import dev.lynith.core.ui.callbacks.impl.WindowResized
import dev.lynith.core.ui.layouts.DefaultLayout
import dev.lynith.core.ui.layouts.Layout
import dev.lynith.core.ui.layouts.LayoutProperties
import dev.lynith.core.ui.styles.ComponentWithChildrenStyles
import java.util.concurrent.CopyOnWriteArrayList

abstract class ComponentWithChildren<C : Component<C, S>, S : ComponentWithChildrenStyles<C, S>> : Component<C, S>() {

    val children: MutableList<Component<*, *>> = CopyOnWriteArrayList()
    var layout: Layout = DefaultLayout()

    fun reposition() {
        layout.position(this)
    }

    fun layout(func: LayoutProperties.() -> Unit): C {
        func(layout.properties)
        return this as C
    }

    override fun postRender(mouseX: Int, mouseY: Int, delta: Float) {
        for (child in children) {
            child.wrappedRender(mouseX, mouseY, delta)
        }

        super.postRender(mouseX, mouseY, delta)
    }

    override fun postInit() {
        super.postInit()

        on<Destroyed> {
            clearChildren()
        }

        for (child in children) {
            child.wrappedInit()
        }

        layout.position(this)
        on<WindowResized> {
            layout.position(this)
        }
    }

    // Children

    open fun child(child: Component<*, *>) {
        children.add(child)
    }

    fun children(vararg children: Component<*, *>) {
        for (c in children) {
            child(c.apply {
                parent = this@ComponentWithChildren
            })
        }
    }

    fun children(children: List<Component<*, *>>) {
        for (c in children) {
            child(c.apply {
                parent = this@ComponentWithChildren
            })
        }
    }

    fun removeChild(child: Component<*, *>) {
        children.remove(child)
    }

    fun clearChildren() {
        children.clear()
    }

}